File: Queue/Queue.swift
// 4.0

internal final class Queue<T> {

  internal var count: Int

  private var items: Array<T?>
  private var first: Int
  private var last: Int

  internal init(count: Int) {
    self.items = [T?](repeating: nil, count: count)
    self.count = 0
    self.first = 0
    self.last = count - 1
  }

  internal func isEmpty() -> Bool {
    return self.count == 0
  }

  internal func isFull() -> Bool {
    return self.count == self.items.count
  }

  internal func peek() throws -> T? {

    guard !self.isEmpty() else {
      throw QueueError.emptyQueue(reason: "Cannot peek at an empty queue!")
    }

    return self.items[self.first]
  }

  internal func dequeue() throws -> T? {

    guard !self.isEmpty() else {
      throw QueueError.emptyQueue(reason: "Cannot dequeue from an empty queue!")
    }

    let item: T? = self.items[self.first]
    self.items[self.first] = nil
    self.first = (self.first + 1) % self.items.count
    self.count -= 1
    return item
  }

  internal func enqueue(item: T?) throws -> Void {

    guard !self.isFull() else {
      throw QueueError.fullQueue(reason: "Cannot enqueue into a full queue!")
    }

    self.last = (self.last + 1) % self.items.count
    self.items[self.last] = item
    self.count += 1
  }

}

internal enum QueueError: Error {
  case emptyQueue(reason: String)
  case fullQueue(reason: String)
}

File: Queue/queue.py
class Queue(object):

    def __init__(self, length):
        self.__items  = [None] * length
        self.__last   = length - 1
        self.__first  = 0
        self.__length = 0

    def __len__(self):
        return self.__length

    @property
    def is_empty(self):
        return self.__length == 0

    @property
    def is_full(self):
        return self.__length == len(self.__items)

    def peek(self):

        if self.is_empty:
            raise IndexError("Cannot peek at an empty queue!")

        return self.__items[self.__first]

    def dequeue(self):

        if self.is_empty:
            raise IndexError("Cannot dequeue from an empty queue!")

        item = self.__items[self.__first]
        self.__items[self.__first] = None
        self.__first = (self.__first + 1) % len(self.__items)
        self.__length -= 1
        return item

    def enqueue(self, item):

        if self.is_full:
            raise IndexError("Cannot enqueue into a full queue!")

        self.__last = (self.__last + 1) % len(self.__items)
        self.__items[self.__last] = item
        self.__length += 1
